HIVE-22167: TIMESTAMP - Backwards incompatible change: Hive 4 reads back binary RCFILE timestamps written by Hive 2.x incorrectly

sercanCyberVision opened the pull request:
**MOTIVATION:**
Hive 4 reads RCFile timestamps written by Hive 2 incorrectly. Please see https://issues.apache.org/jira/browse/HIVE-22167 for more details.
 
**SOLUTION:**
A new property has been added to switch between current and legacy interpretation.
```
<property>
  <name>hive.rcfile.timestamp.legacy.conversion</name>
  <value>true</value>
  <description>Whether to convert legacy RCFile timestamps (Hive 2.x format)</description>
</property>
```

**CHECKS:** 
```
hive> set hive.rcfile.timestamp.legacy.conversion;
hive.rcfile.timestamp.legacy.conversion=false --------->> DEFAULT
hive> SELECT * FROM t;
OK
2025-07-01 12:48:15.01295 -------->> Incorrect value

--SET TO TRUE:

hive> set hive.rcfile.timestamp.legacy.conversion=true;
hive> SELECT * FROM t;
OK
2025-07-01 14:48:15.01295 -------->> Correct value
Time taken: 0.222 seconds, Fetched: 1 row(s)
hive>
```

deniskuzZ commented:
@sercanCyberVision, Hive 3 or Hive 4? Note that Hive 3 is officially EOL.

deniskuzZ commented on serde/src/java/org/apache/hadoop/hive/serde2/lazybinary/LazyBinaryTimestamp.java:
```diff
@@ -32,6 +37,13 @@ public class LazyBinaryTimestamp extends
   static final Logger LOG = LoggerFactory.getLogger(LazyBinaryTimestamp.class);
 
+  // Legacy mode flag
+  private static boolean legacyModeEnabled = false;
```
that should be a reader property, see for example `ParquetRecordReaderBase.legacyConversionEnabled`

sercanCyberVision replied:
Thank you!

sercanCyberVision commented:
> @sercanCyberVision, Hive 3 or Hive 4? Note that Hive 3 is officially EOL.

@deniskuzZ both Hive 3 and Hive 4 have the same behavior. 
This issue was already addressed for the Parquet and Avro formats in both versions — please see:
- https://issues.apache.org/jira/browse/HIVE-21290
- https://issues.apache.org/jira/browse/HIVE-21291

deniskuzZ commented:
@sercanCyberVision, please add the tests to validate the legacy functionality

sercanCyberVision commented:
@deniskuzZ, I have corrected the description and Jira title. 
Thank you for the patch!
Now working on creating a test

deniskuzZ commented on serde/src/java/org/apache/hadoop/hive/serde2/lazybinary/LazyBinaryTimestamp.java:
```diff
@@ -52,6 +55,28 @@ public class LazyBinaryTimestamp extends
+      data.set(bytes.getData(), start);
+
+      // Convert to java.sql.Timestamp
+      Timestamp ts = data.getTimestamp();
```
if that is already handled by other file formats, can't we reuse/extract the conversion logic into util class?

sercanCyberVision commented:
@deniskuzZ 
I'm encountering an issue with timestamp conversion in legacy mode.
When timestamps are read from a table (e.g., during a join) and rewritten, they undergo multiple unintended conversions.

I added ZoneID to Timestamp class to use it as condition (if zoneIDs are the same, do not convert anymore), but I am losing the ZoneID at some point (I assume during the serialization), and it keeps re-converting.

For example:
```
SET hive.rcfile.timestamp.legacy.conversion=true;
CREATE TABLE t1 (c1 TIMESTAMP, c2 TIMESTAMP);
CREATE TABLE t2 (c1 TIMESTAMP, c2 TIMESTAMP) STORED AS RCFILE;
INSERT INTO t1 VALUES ('2025-07-29 00:00:01','2025-07-29 00:00:01');
INSERT INTO t2 VALUES ('2025-07-29 00:00:01','2025-07-29 00:00:01');


INSERT OVERWRITE TABLE t2
SELECT
    ext.c1,
    mng.c2
FROM
    t1 AS ext
  FULL OUTER JOIN
    t2 AS mng
  ON
    1 = 1;


SELECT * FROM t1;
SELECT * FROM t2;
```
returns;
```
hive> SELECT * FROM t1;
OK
2025-07-29 00:00:01	2025-07-29 00:00:01
Time taken: 0.169 seconds, Fetched: 1 row(s)
hive> SELECT * FROM t2;
OK
2025-07-29 04:00:01	2025-07-29 06:00:01
Time taken: 0.165 seconds, Fetched: 1 row(s)
hive>
```
Could you please give me any suggestion?
